### ISSUE TITLE
^M in DICOM Files

### ISSUE THREAD

**Issue by bkraft4257** [nipy/heudiconv#137]
I have come across a unique problem running heudiconv on DICOM images created on a Siemens scanner upgraded to E11C.  This problem is created by a "corrupt" DICOM image and not a bug in the heudiconv code.  However, I have now encountered the problem twice on  DICOM images created by two different pulses sequence (multiband EPI provided by Michigan and a product TOF angio product sequence).  Although I don't expect you to address this issue I thought you might want to be at least aware of it. 

These DICOM images have a ^M in their headers.  These ^M are converted when the DICOM images are parsed and are placed in the JSON file, which corrupts the JSON files.   Here is the traceback

```
Traceback (most recent call last):
  File "/cenc/software/heudiconv/python/heudiconv/bin/heudiconv", line 2273, in <module>
    main()
  File "/cenc/software/heudiconv/python/heudiconv/bin/heudiconv", line 2265, in main
    return _main(args)
  File "/cenc/software/heudiconv/python/heudiconv/bin/heudiconv", line 2121, in _main
    min_meta=args.minmeta)
  File "/cenc/software/heudiconv/python/heudiconv/bin/heudiconv", line 1474, in convert_dicoms
    min_meta=min_meta)
  File "/cenc/software/heudiconv/python/heudiconv/bin/heudiconv", line 1021, in convert
    tuneup_bids_json_files(outname_bids_files)
  File "/cenc/software/heudiconv/python/heudiconv/bin/heudiconv", line 1200, in tuneup_bids_json_files
    json_ = json.load(open(jsonfile))
  File "/opt/anaconda2/lib/python2.7/json/__init__.py", line 291, in load
    **kw)
  File "/opt/anaconda2/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/opt/anaconda2/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/opt/anaconda2/lib/python2.7/json/decoder.py", line 380, in raw_decode
    obj, end = self.scan_once(s, idx)
ValueError: Invalid control character at: line 25 column 48 (char 887)
```

Here is a portion of the JSON file(cut and pasted when viewing the file in emacs)

        "SeriesNumber": 15,
        "AcquisitionTime": "13:31:48.057500",
        "AcquisitionNumber": 1,
        "ImageComments": "RR_959_+_-_63;_66_heartbeats^M
        "SliceThickness": 8,
        "SAR": 0.251429,
        "EchoTime": 0.00386,
        "RepetitionTime": 0.03972,

You can see the ^M in the "ImageComments".  Since the JSON file is corrupt heudiconv fails.  This problem is clearly an issue with the DICOM file with the Image Comments field in both cases. You can read the DICOM file in with pydicom and see the offending ImageComments field for phcon_00001.DCM

```
>>> import dicom
>>> ds = dicom.read_file("000001.DCM")
>>> ds.ImageComments
'RR 959 +/- 63; 66 heartbeats\r\nE11C UPGRADE'
```

Here is the same dicom field for mbep_000001.DCM

```
>>> ds = dicom.read_file("mbepi_000001.DCM")
>>> ds.ImageComments
'Unaliased MB8/PE3/LB\r\nE11C UPGRADE'
```

I think the problem is with the \r\nE11C UPGRADE because it appears in both fields. The E11C_UPGRADE is put on the next line of the JSON file.  

I have hacked (and I do mean hack) where I perform sed on the JSON files before they are loaded by heudiconv.  This works but I have had to modify my sed for each instance of the DICOM error I find.  I am looking into trying to remove the ^M from the DICOM images before I call heudiconv.  If others have other solutions I would be like to hear them.

I hope this information helps,
 
Bob

P.S. heudiconv is a great product. Thank you for creating, maintaining and sharing it.

P.P.S. Sorry for not posting the TraceBack. I hit the Enter key a little too early. 

[dicom_with_control_m.zip](https://github.com/nipy/heudiconv/files/1657745/dicom_with_control_m.zip)

**Comment by yarikoptic** [comment]
traceback got lost ;)  but I think it is the same which we observed but it was left a mystery since dcm2niix's output did not have them (IIRC it was ImageComments field), and I could have sworn an oath that conversion worked before (so something changed but we could not identify what)

**Comment by yarikoptic** [comment]
ok -- in our case it is due to those MoCoSeries
```shell
$> dcm2niix -b y -z i -x n -t n -m n -f func -o /tmp/out -s n -v n dartmouth-phantoms/bids_test4-20161014/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409265217734302222.dcm
Chris Rorden's dcm2niiX version v1.0.20171103 (OpenJPEG build) GCC7.2.1 (64-bit Linux)
Found 8 DICOM image(s)
slices stacked despite varying acquisition numbers (if this is not desired please recompile)
Slice positions repeated, but number of slices (2) not divisible by number of repeats (2): missing images?
 Warning: all images appear to be a single slice - please check slice/vector orientation
Convert 2 DICOM as /tmp/out/func (64x64x1x2)
Warning: Check that 2D images are not mirrored.
slices stacked despite varying acquisition numbers (if this is not desired please recompile)
Slice positions repeated, but number of slices (2) not divisible by number of repeats (2): missing images?
 Warning: all images appear to be a single slice - please check slice/vector orientation
Convert 2 DICOM as /tmp/out/funca (64x64x1x2)
Warning: Check that 2D images are not mirrored.
slices stacked despite varying acquisition numbers (if this is not desired please recompile)
Slice positions repeated, but number of slices (2) not divisible by number of repeats (2): missing images?
 Warning: all images appear to be a single slice - please check slice/vector orientation
Convert 2 DICOM as /tmp/out/funcb (64x64x1x2)
Warning: Check that 2D images are not mirrored.
slices stacked despite varying acquisition numbers (if this is not desired please recompile)
Slice positions repeated, but number of slices (2) not divisible by number of repeats (2): missing images?
 Warning: all images appear to be a single slice - please check slice/vector orientation
Convert 2 DICOM as /tmp/out/funcc (64x64x1x2)
Warning: Check that 2D images are not mirrored.
Conversion required 0.008434 seconds (0.007848 for core code).

$> grep ImageComm /tmp/out/*json                          
/tmp/out/funca.json:	"ImageComments": "Reference_volume_for_motion_correction.
/tmp/out/funcb.json:	"ImageComments": "Reference_volume_for_motion_correction.",
/tmp/out/funcc.json:	"ImageComments": "Reference_volume_for_motion_correction.
/tmp/out/func.json:	"ImageComments": "Reference_volume_for_motion_correction.",
(dev) 2 22917.....................................:Tue 23 Jan 2018 04:49:05 PM EST:.
hopa:~/picts/mris
$> grep -A1 ImageComm /tmp/out/*json
/tmp/out/funca.json:	"ImageComments": "Reference_volume_for_motion_correction.
/tmp/out/funca.json-image_comment_2",
--
/tmp/out/funcb.json:	"ImageComments": "Reference_volume_for_motion_correction.",
/tmp/out/funcb.json-	"SliceThickness": 3.5,
--
/tmp/out/funcc.json:	"ImageComments": "Reference_volume_for_motion_correction.
/tmp/out/funcc.json-image_comment_2",
--
/tmp/out/func.json:	"ImageComments": "Reference_volume_for_motion_correction.",
/tmp/out/func.json-	"SliceThickness": 3.5,

$> dcm2niix --version
Chris Rorden's dcm2niiX version v1.0.20171103 (OpenJPEG build) GCC7.2.1 (64-bit Linux)
 Error: invalid option '--version (null)'
Example output filename: '/func.nii.gz'
```

**Comment by yarikoptic** [comment]
aha - my issue was addressed in a recent dcm2niix (built 1.0.20171215+git8-g82b9210-1 locally, will do now for neurodebian). Here is a full log FWIW;
```shell
$> datalad install ///dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-1/
[INFO   ] Cloning http://datasets.datalad.org/dicoms/dartmouth-phantoms/bids_test4-20161014/phantom-1/ to '/tmp/phantom-1' 
install(ok): /tmp/phantom-1 (dataset)
(dev) 2 23057.....................................:Tue 23 Jan 2018 11:44:24 PM EST:.
hopa:/tmp
$> rm -rf out; mkdir out; 
(dev) 2 23058.....................................:Tue 23 Jan 2018 11:44:34 PM EST:.
hopa:/tmp
$> datalad get phantom-1/MoCoSeries                                                                                                                  
[INFO   ] Actually getting 8 files 
get(ok): /tmp/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409265217734302222.dcm (file)                                                                                            
get(ok): /tmp/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409265385049702228.dcm (file)
get(ok): /tmp/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409270575369702242.dcm (file)
get(ok): /tmp/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409270823880402248.dcm (file)
get(ok): /tmp/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409272069411002262.dcm (file)
get(ok): /tmp/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409272317892602268.dcm (file)
get(ok): /tmp/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409273584407202282.dcm (file)
get(ok): /tmp/phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409273832740502288.dcm (file)
get(ok): /tmp/phantom-1/MoCoSeries (directory)
action summary:
  get (ok: 9)
(dev) 2 23059.....................................:Tue 23 Jan 2018 11:44:59 PM EST:.
hopa:/tmp
$> dcm2niix -b y -z i -x n -t n -m n -f func -o out -s n -v n phantom-1/MoCoSeries/1.3.12.2.1107.5.2.43.66112.2016101409265217734302222.dcm
Chris Rorden's dcm2niiX version v1.0.20171215 (OpenJPEG build) GCC7.2.0 (64-bit Linux)
Found 8 DICOM image(s)
slices stacked despite varying acquisition numbers (if this is not desired please recompile)
Slice positions repeated, but number of slices (2) not divisible by number of repeats (2): missing images?
 Warning: all images appear to be a single slice - please check slice/vector orientation
Convert 2 DICOM as out/func (64x64x1x2)
Warning: Check that 2D images are not mirrored.
slices stacked despite varying acquisition numbers (if this is not desired please recompile)
Slice positions repeated, but number of slices (2) not divisible by number of repeats (2): missing images?
 Warning: all images appear to be a single slice - please check slice/vector orientation
Convert 2 DICOM as out/funca (64x64x1x2)
Warning: Check that 2D images are not mirrored.
slices stacked despite varying acquisition numbers (if this is not desired please recompile)
Slice positions repeated, but number of slices (2) not divisible by number of repeats (2): missing images?
 Warning: all images appear to be a single slice - please check slice/vector orientation
Convert 2 DICOM as out/funcb (64x64x1x2)
Warning: Check that 2D images are not mirrored.
slices stacked despite varying acquisition numbers (if this is not desired please recompile)
Slice positions repeated, but number of slices (2) not divisible by number of repeats (2): missing images?
 Warning: all images appear to be a single slice - please check slice/vector orientation
Convert 2 DICOM as out/funcc (64x64x1x2)
Warning: Check that 2D images are not mirrored.
Conversion required 0.005986 seconds (0.005955 for core code).
(dev) 2 23060.....................................:Tue 23 Jan 2018 11:45:02 PM EST:.
hopa:/tmp
$> grep ImageComm out/*json                                  
out/funca.json:	"ImageComments": "Reference_volume_for_motion_correction.",
out/funcb.json:	"ImageComments": "Reference_volume_for_motion_correction._image_comment_2",
out/funcc.json:	"ImageComments": "Reference_volume_for_motion_correction._image_comment_2",
out/func.json:	"ImageComments": "Reference_volume_for_motion_correction.",
```

**Comment by yarikoptic** [comment]
so, as it was fixed in the recent dcm2niix (please upgrade), closing this issue
